### DISCUSSION TITLE
Self Hosting Athens as a web app?

### DISCUSSION THREAD

**Discussion by ghost** [athensresearch/athens#737] [Question / Answer]
Can Athens be self-hosted as a web app?  More or less as the online demo site, but with the ability to save (of course).  I have the need to access Athens from multiple workstations and having it as a local app isn't the best fit.  Any guidance in how to do this would be appreciated.  I came across the HN post just as I was about to subscribe to Roam so this is very exciting!

**Comment by jsmorabito** [comment]
We are planning to have a web app but I can't provide any estimate on the timeline 😞 If you'd like to follow along with development we are actively working here on github and in our discord https://discord.gg/as9h8yHNfD News about progress on the web app would likely be sent on our #announcements channel on discord which you can subscribe to. Alternatively, are there any other ways that you'd like to be contacted about the projects progress?

**Comment by Limezy** [comment]
Definitely waiting for that one. And please, if you can not just only a docker container !
When it'll be released I'll try to package it for Yunohost https://github.com/YunoHost/yunohost

**Reply by tomisme** [reply]
Oh wow a yunohost package would be fantastic!

**Reply by Limezy** [reply]
Yep 👍 . But before working on it, let's wait for Athens team to release their web app. My guess is that it shouldn't be so long since they already run it as a demo on their website

**Reply by jefftangx** [reply]
The backend server is very close :)

**Reply by Limezy** [reply]
@tomisme FYI I nearly managed to install Athens as a webapp on my Yunohost test instance.
I was blocked with a final error I can't solve, I think it's because of some recent updates that were made in the code.
However, the good news is that I'm now 99% sure there will be no problem to build an Athens yunohost package as an alternative way to self-host it instead of using the (future) docker package

**Comment by Limezy** [comment]
Hi, because of our discussion with @tangjeff0 on discussion https://github.com/athensresearch/athens/discussions/986, I've tried to make ~~a little summary~~ a gigantic post about what could be the uses cases for Athens users. For all the below use cases, I've assumed that the Mobile app exists.

### First trial to list out all use cases

For me there are three main parameters :
**1. First parameter is where the Athens user will run Athens software (frontend):** 
* Could be his browser, Athens Research hosted
* Could be in his browser, served by his "own server" (Deta, Yunohost, others...)
* Could be in a dedicated app (Desktop or Mobile)

**2. Second parameter is where that Athens user will store his database :** 
* Could be in an Athens backend, Athens Research hosted
* Could be in an Athens backend, served by his "own server" (Deta, Yunohost, others...)
* Could be in his own "cloud" (Google Drive, Dropbox, OneDrive, Nextcloud...)
* Could be purely locally

**3. Third parameter is whether that user needs Real Time Collaboration or not**
* Could be yes
* Could be yes and no (for instance personal graph doesn't need it, professional graph needs it)
* Could be no

This makes 36 different use cases 😅 so let's simplify things.

### Simplification of use cases
Let's remove the third parameter, as it's quite easy to connect with the second parameter.

**1. First parameter is where the Athens user will run Athens software (frontend):** 
* Could be his browser, Athens Research hosted
* Could be in his browser, served by his "own server" (Deta, Yunohost, others...)
* Could be in a dedicated app (Desktop or Mobile)

**2. Second parameter is where that Athens user will store his database :** 
* Could be in an Athens backend, Athens Research hosted [RTC ✅]
* Could be in an Athens backend, served by his "own server" (Deta, Yunohost, others...) [RTC ✅]
* Could be in his own "cloud" (Google Drive, Dropbox, OneDrive, Nextcloud...) [RTC ❌]
* Could be purely locally [RTC ❌]

That leaves us with **12 use cases** "only", but some or them are not to be considered, as you'll see below.

### Summary of the use cases
![image](https://user-images.githubusercontent.com/24638389/115437553-e646be00-a23e-11eb-9890-31f3faaae5a8.png)

In the end, we end up with only 5 distinct use cases.

### Why are all these 5 use cases legit ?

1. I think it's legitimate that somebody needs to run Athens on a browser, and not only in a app. Simple example, the guy who uses Athens but whose employer doesn't allow to install any app on his computer. This is also why webmails have become so popular
2. But it's also legitimate to want to use Athens in a native app, which will always provide a better experience
3. For somebody who doesn't care about RTC, and that already have all his files on Dropbox, Google Drive or whatsoever, seems legitimate if he just wants to keep his notes safely at the same place (knowing that there is a big chance he won't be able to sync with his mobile)
4. I think it's legitimate that somebody wants to serve Athens (frontend or backend) from his own hardware. For safety or privacy reasons, for the sake of having all your digital life in your own server... 
5. I also think it's legitimate that somebody wants to use Athens in "Software as a Service = (SaaS) mode (frontend and/or backend), for confort, added value, because his company has signed a contract with Athens Research, or just because it's an easy way to support the project !

### Last considerations about these use cases : 
* **Case A** : obviously, this is the pristine Athens experience. RTC compatible, mobile App compatible, full SaaS but with native apps environment for confort etc... 
* **Case B** : same first class experience, RTC compatible, mobile App compatible, but you have to go through the hassle of building your own architecture, which would lead most of users to **Case A**. I don't exactly know how the Athens team could deal with third parties (Deta.space, Heroku ?) that could invoice for the same service as **Case A**. Other third parties, like Dockers, Yunohost, Cloudron, Sandstorm.io... could help reducing the effort of building your own architecture, but I don't think that will appeal to anybody that doesn't already have his own setup. And I guess most of **Case B** users are already strong Open Source believers and will support the project anyway (at least, that would be my case)
* **Case C** : I suspect that case to be the most common today. The mobile app, if it's not compatible with the standard cloud services, may drive a big part of that user base to **Case A**. And even if the mobile app was compatible with the standard cloud services, that won't bring RTC to them, so again, another probable shift to **Case A**.
* **Case D** : I'm not sure Case D is an option for Athens team, so let's not spend time on it
* **Case E** : Probably a Athens-newby guy that will quickly move to **Case C**, before moving to **Case A** 😄

**Reply by jefftangx** [reply]
Wow, thank you for this comprehensive analysis @Limezy -- I actually don't have too much to add!

To summarize and add some comments:

* **Case A**: mature SaaS, comparable to Notion or Google Docs 
* **Case B**: mature self-hosted/on-premise, comparable to GitLab or Mattermost
* **Case C**: local-only application w/ cloud sync but w/o real collaboration, probably where most of our users are today, comparable to Joplin or org-mode
* **Case D**: LogSeq stores data in the browser using IndexedDB. We will eventually use IndexedDB, but as a cache (via Datahike), not as a persistent store. The reason we wouldn't use IDB as a persistent store is that a server is probably still necessary for real collaboration. (Runs away from blockchain/crypto people.)
* **Case E**: I think both Case E and Case C will be around for a while, at least until we have stronger security and privacy features like e2ee. Even with a mature SaaS, I think there's still some cases where a user might want a local-only db. Besides, with #882, it should also be trivially easy to switch between local and remote DBs!
